[MRG+3] FEA Add PolynomialCountSketch to Kernel Approximation module

lopeLH opened the pull request:
This PR adds the **Tensor Sketch** [1] algorithm for polynomial kernel feature map approximation to the Kernel Approximation module. 

Tensor Sketch is a well established method for kernel feature map approximation, which has been broadly applied in the literature. For instance, it has recently gained a lot of popularity to accelerate certain bilinear models [2]. While the current kernel approximation module contains various kernel approximation methods, polynomial kernels are missing, so including TensorSketch completes the functionality of this module by providing an efficient and data-independent polynomial kernel approximation technique. 

The PR contains the implementation of the algorithm, the corresponding tests, an example script, and a description of the algorithm in the documentation page of the kernel approximation module. This implementation has been tested to produce the same results as the original matlab implementation provided by the author of the algorithm [1].

[1] [Pham, N., & Pagh, R. (2013, August). Fast and scalable polynomial kernels via explicit feature maps. In Proceedings of the 19th ACM SIGKDD international conference on Knowledge discovery and data mining (pp. 239-247). ACM.](https://scholar.google.es/scholar?hl=es&as_sdt=0%2C5&q=Fast+and+scalable+polynomial+kernels+via+explicit+feature+maps&btnG=)

[2] [Gao, Y., Beijbom, O., Zhang, N., & Darrell, T. (2016). Compact bilinear pooling. In Proceedings of the IEEE conference on computer vision and pattern recognition (pp. 317-326).](https://scholar.google.es/scholar?hl=es&as_sdt=0%2C5&q=Compact+bilinear+pooling%2C+Y.+Gao&btnG=)
_____
Work for follow-up PR:

- [ ] @rth: This can be a follow up issue/PR, and would need double checking but since the count_sketches input is real you can likely use rfft and irfft witch would be faster.

- [ ] @rth: The issue is that calling fit twice would produce a different seed and therefore a different result, since RandomState instance is mutable. In #14605 for a similar use case, we added a seed variable in transform, but I'm not particularly happy with that outcome either. This is probably fine as is, we would just have to address this globally at some point in #14042

- [ ] @rth: It could be worth considering whether it would make sense thesholding (in the above example 1e-15 would be ok as a threshold) and converting back to sparse. Though the intermediary step with the FFT would still be dense with the associated memory requirements, maybe it could be worth chunking with respect to n_samples not sure (https://github.com/scikit-learn/scikit-learn/pull/13003#issuecomment-673424927).

lopeLH committed:
Added Tensor Sketch algorithm

Added Tensor Sketch algorithm in the kernel approximation module.

lopeLH committed:
some cleaning

lopeLH committed:
code style

removed or added white spaces where needed.

lopeLH committed:
more style edits

lopeLH committed:
style edits

lopeLH committed:
style edits

lopeLH committed:
style edits

lopeLH committed:
add tets for Tensor Sketch

lopeLH committed:
style edits

lopeLH committed:
style edits

lopeLH committed:
style edits

lopeLH committed:
style edits

lopeLH committed:
style edits

lopeLH committed:
style edits

lopeLH committed:
style edits

lopeLH committed:
more and more style edits...

lopeLH committed:
cleaning

lopeLH committed:
-

lopeLH committed:
style edits

lopeLH committed:
Added an example and optimized implementation

lopeLH committed:
style edits

lopeLH committed:
style edits

lopeLH committed:
style edits

lopeLH committed:
fix some comments

lopeLH committed:
added entry in documentation

lopeLH committed:
style edits

lopeLH committed:
style edits

lopeLH committed:
style edits

lopeLH committed:
add support for non-homogeneous poly kernels

This is, polynomial kernels with coef0 != 0

GaelVaroquaux commented:
This definitely looks interesting. The question is whether the paper has enough citations to be eligible for inclusion in scikit-learn. 

It has 150 citations, so it is borderline. It's unclear to me if the contribution should go here, it in scikit-learn-contrib

lopeLH commented:
An argument in favor of inclusion in scikit-learn is that, while the original paper [1] has 150 citations, some recent papers which basically apply this algorithm in a different context have 203 and 347 citations respectively [2, 3]. 

There is also the argument of completeness for the Kernel Approximation module, which currently hosts data-independent approximation methods for some of the most popular kernels (RBF, Chi Squared) but not for Polynomial Kernels.

[1] Pham, N., & Pagh, R. (2013, August). Fast and scalable polynomial kernels via explicit feature maps. In Proceedings of the 19th ACM SIGKDD international conference on Knowledge discovery and data mining (pp. 239-247). ACM.
[2] Gao, Y., Beijbom, O., Zhang, N., & Darrell, T. (2016). Compact bilinear pooling. In Proceedings of the IEEE conference on computer vision and pattern recognition (pp. 317-326).
[3] Fukui, A., Park, D. H., Yang, D., Rohrbach, A., Darrell, T., & Rohrbach, M. (2016). Multimodal compact bilinear pooling for visual question answering and visual grounding. arXiv preprint arXiv:1606.01847.

amueller commented:
If we include it, should it be called something that shows the usage instead of the algorithm? Not sure if calling RBFSampler RBFSampler was the best, but TensorSketch is not very discoverable for someone trying to approximate Polynomial kernels.

lopeLH commented:
Perhaps it should be called something like PolynomialSampler, which would be more descriptive and consistent with the rest of the approximators in the module.

lopeLH commented:
Any news about this @GaelVaroquaux @amueller? Should I go ahead and change the name of algorithm to PolynomialSampler, or you have another name in mind?

agramfort commented:
it would be great to have a benchmark that demonstrates the benefit on a couple of real datasets.

my 2c

GaelVaroquaux commented:
Looking at the example, it seems to me that TensorSketch is slower and less performant than Nystrom:
https://43991-843222-gh.circle-artifacts.com/0/doc/auto_examples/plot_polynomial_kernel_approximation.html

Am I reading them wrong?

lopeLH commented:
My bad. The example is comparing transform times only, while the major advantage of TensorSketch is in the fit phase, since it only initializes some random matrices whereas Nystrom performs SVD on a (_n_components_ x _n_components_) matrix.

In my machine TensorSketch was also faster to transform samples though, but it seems like _k_ needs to get bigger before Nystrom's O(n(dk+k^2)) transform time surpasses TensorSketch's O(n(d+k log k)) transform time in the server machine. 

The slightly superior accuracy of Nystrom was expected, as it is actually using training data while TensorSketch is just a data-independent randomized feature map.

I will update the example to compare fit_transform times rather than transform times, which might be more relevant in this case.  I am also working on some examples with real-world data, but that might take a bit longer.

Sorry about that.

lopeLH committed:
Measure fit_transform time in the example

Measure fit_transform time in the example instead of just measuring
transform times.

lopeLH commented:
As expected, when fit times are also considered TensorSketch scales much better:

<img src="https://48476-843222-gh.circle-artifacts.com/0/doc/_images/sphx_glr_plot_polynomial_kernel_approximation_002.png"></img>

GaelVaroquaux commented:
Very nice!

agramfort commented:
I am also inclined to accept this but the example currently looks more like a benchmark script than an example. I would put the current code in scikit-learn/benchmarks folder and integrate this new kernel approximation technique to an existing example using nystrom form example.

GaelVaroquaux commented:
OK. I think that we have a consensus to include this hurray!

+1 on renaming to PolynomialSampler

+1 on making this example more oriented toward end users. That said, the current Nystrom example is strongly oriented toward the RBF kernel. Maybe it would be useful to actually have an example on scalable learning of polynomial regressions that introduces polynomial sampler and SGDClassifier. This would probably make the feature contributed here discoverable by users.

Thanks for the contribution @lopeLH . This is exciting!

lopeLH committed:
Change the name of TensorSketch to PolynomialSampler

Change the name of TensorSketch to PolynomialSampler to make it more
discoverable.

lopeLH committed:
fix typos in the example

lopeLH committed:
import with the new name in the tests

lopeLH committed:
update doc and fix error in comments

agramfort commented on doc/modules/kernel_approximation.rst:
```diff
@@ -154,6 +154,24 @@ above for the :class:`RBFSampler`. The only difference is in the free
+:math:`n` is the number of samples, :math:`d` the dimension of input samples
+and :math:`k` the desired output dimension, determined by (``n_components``). 
+
+.. _tensor_sketch_kernel_approx:
```
this narrative doc here should point the example that demos the usage with the figure produced by it.
can you take care of it?

also I would really suggest to simplify the example so it demos the usage and we keep the full benchmark for the benchmarks folder here: https://github.com/scikit-learn/scikit-learn/tree/master/benchmarks

lopeLH replied:
Yes, sure.

I will move the current example to the benchmarks folder. I am working on a simpler example with a real-world dataset, as you suggested.

lopeLH committed:
style edits

lopeLH committed:
accept sparse matrix input

lopeLH committed:
move old example to benchmarks and add new example

Move old example to benchmark dataset. Add a new, simpler example.
Reference example form documentation.

lopeLH commented:
I made the following changes:

1. Added support for sparse input matrices.
2. Moved old example to the benchmarks folder.
3. Added a new example which illustrates the use of PolynomialSampler in the scalable approximation of polynomial kernels for classification. This replicates the results of the original paper on the Covtype dataset. Here I used LinearSVC instead of SGDClassifier for consistency with the original paper, but I can change that if you think is better. 
4. Referenced the new example from the narrative documentation. 

Thank you for your time and patience!

lopeLH committed:
fix typo

fixes a typo in the comments of
bench_plot_polynomial_kernel_approximation.py

lopeLH commented:
I don't know why [ci/circleci: doc](https://circleci.com/gh/scikit-learn/scikit-learn/51657?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) is failing... It gives the following error when executing the example.

Makefile:43: recipe for target 'html' failed
make: *** [html] Error 137

Is it because running the example requires too much resources? should I make it less resource-demanding?

jnothman commented:
Yes, likely ran out of memory.

lopeLH committed:
make example less memory-demanding

lopeLH committed:
simplify example even more

lopeLH committed:
more edits in the example

lopeLH committed:
improve coverage

test that PolynomialSampler works with sparse input data

lopeLH committed:
fix typo in comments

lopeLH commented:
All checks passing. Hurray!

I think there is nothing left for me to do. If you think additional changes are needed, please let me know.

lopeLH commented:
Any news about this?

lopeLH committed:
Merge branch 'master' into add-tensorSketch

lopeLH committed:
fix lint error on kernel approximation tests

lopeLH committed:
fix example format to pass checks

lopeLH committed:
fix deprecated use of check_is_fitted

lopeLH committed:
fix kernel approximation minimal example style

lopeLH committed:
try to fix example syntax

lopeLH commented:
All test passing again, hurray! 

Waiting for the reviews :) @GaelVaroquaux @amueller @agramfort @jnothman

lopeLH committed:
Merge branch 'master' into add-tensorSketch

GaelVaroquaux commented:
Hi @lopeLH 

Sorry we are terribly unresponsive on this PR. I want to try to make a push and get it merged. But it's always a big decision on our side.

To help me review it, could you paste here the figures generated by the benchmark script.

Thanks a lot for your thoroughness and your patience.

GaelVaroquaux commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,96 @@
+"""
```
I think that this example would be great improved by adding a few separators and text, to make a "notebook style example". 
See https://sphinx-gallery.github.io/stable/tutorials/index.html

It's just a question of adding a few separators to insert titles, have "prints" display there results next to where there are in the code, and to turn some comments in text blocks.

lopeLH replied:
Done!

lopeLH committed:
[doc] notebook style example

Add separators to the example for PolynomialSampler to make it
notebook-style.

lopeLH commented:
No problem @GaelVaroquaux 👍. Following your requests:

- [ ] Modify example with text and separators to make it more notebook-style.
- [ ] Paste benchmark results here (see below).

So the benchmark script is generating two plots. The first one shows the accuracy of a linear SVM classifier (dashed black), a polynomial kernel SVM (dashed red), and how `Nystroem` and `PolynomialSampler` methods can approximate the performance of the kernel SVM. The message here is that, for a sufficiently large number of generated features (`n_components`), **both methods achieve a good approximation of the accuracy of the kernel-SVM, but using a linear classifier!** `Nystroem` seems to perform slightly better than `PolynomialSampler`, but this was expected since `Nystroem` is actually using training data while `PolynomialSampler` is just a data-independent randomized feature map.

![image](https://user-images.githubusercontent.com/7440734/74884140-c5e94e00-5372-11ea-91d5-93b9ec5c0b2a.png)

The second plot **illustrates the remarkable scalability of `PolynomialSampler` when we consider `fit_tranform()` time as a function of the number of generated features (`n_components`).** This is important because, as shown in the previous plot, the number of features has a direct impact in the accuracy.

![image](https://user-images.githubusercontent.com/7440734/74884717-1e6d1b00-5374-11ea-8839-0621aa2fee3a.png)

So the message here is, if you need fast, super scalable and versatile polynomial features, `PolynomialSampler` is the way to go.

lopeLH commented:
@GaelVaroquaux @agramfort @amueller In spite of Covid, I am available to put some work on this PR if needed :muscle:

lorentzenchr commented:
@lopeLH Could you "register" the new `PolynomialSampler` into `doc/modules/classes.rst` in order for the user guide links to work correctly?

lopeLH committed:
[doc] register new class in classes.rst

lopeLH commented:
@lorentzenchr Done! :rocket:

lorentzenchr commented on doc/modules/classes.rst:
```diff
@@ -710,6 +710,7 @@ Plotting
    kernel_approximation.Nystroem
    kernel_approximation.RBFSampler
    kernel_approximation.SkewedChi2Sampler
+   kernel_approximation.PolynomialSampler
```
Please in alphabetical order.

lopeLH replied:
done.

lorentzenchr reviewed (commented):
@lopeLH I personally like sketching, so let's get this PR ready towards final reviews.
General feedback for the example:
- It's always nice to have a plot. Is there something worth plotting in this case?
- You use several transformers. Using pipelines might improve this example.

lopeLH committed:
Apply suggestions from lorentzenchr's code review

Co-Authored-By: Christian Lorentzen <lorentzen.ch@googlemail.com>

lopeLH committed:
[doc] Modify TensorSketch description in the documentation

TomDLT commented on sklearn/kernel_approximation.py:
```diff
@@ -22,7 +24,158 @@
+            raise ValueError("Number of features of test samples does not"
+                             " match that of training samples.")
+
+        Ps = np.zeros((X.shape[0], self.degree, self.n_components))
```
What is Ps ? We try to avoid variable names without meaning. If you think it is still better this way, you could add a comment to state what it means.

lopeLH replied:
- [x] Give intermediate variables more meaningful names.

TomDLT commented on sklearn/kernel_approximation.py:
```diff
@@ -22,7 +24,158 @@
+                    iHashBit = self.bitHash_[d, j]
+                    Ps[:, d, iHashIndex] += iHashBit * X[:, j]
+
+        Ps = fftpack.fft(Ps, axis=2, overwrite_x=True)
```
I would avoid reusing the same variable name `Ps`, for the sake of clarity.

lopeLH replied:
- [ ] Give intermediate variables more meaningful names.

TomDLT commented on sklearn/tests/test_kernel_approximation.py:
```diff
@@ -20,6 +21,54 @@
+    kernel_approx = np.dot(X_trans, Y_trans.T)
+
+    error = kernel - kernel_approx
+    assert np.abs(np.mean(error)) <= 0.01  # close to unbiased
```
We usually use `assert_array_almost_equal(kernel, kernel_approx)`, would it be sufficient ? or is the approximation too coarse ?

lopeLH replied:
I am afraid so. I borrowed this from `test_rbf_sampler( )`.

lopeLH commented on doc/modules/kernel_approximation.rst:
```diff
@@ -149,6 +149,28 @@ above for the :class:`RBFSampler`. The only difference is in the free
+--------------------------------------------------------
+
+The TensorSketch [PP2013]_ method, as implemented in :class:`PolynomialSampler`, is a
+scalable, data-independent method for polynomial kernel approximation. This method
```
It also needs no information about the distribution of training data samples. So I will replace that with `data distribution-independent`. Is that okay?

lorentzenchr replied:
Actually, after having read more, "data-independent" is quite a good description. Maybe you just explain that a bit further by saying something like "data-independent, i.e. only depends on the number of feature".

lorentzenchr replied:
Sorry for the shilly-shallying.

lopeLH commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,119 @@
+:class:`PolynomialSampler`).
+
+First, we compute the accuracy of a linear classifier on the original
+features. Then, we train linear classifiers on different numbers of features
```
Yes. Clarified that.

lopeLH commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,119 @@
+
+lsvm = LinearSVC()
+lsvm.fit(X_train, Y_train)
+print("Linear SVM score on raw features: %.2f %%"
+      % (100*lsvm.score(X_test, Y_test)))
```
I'll go with the f-string :+1:

lopeLH commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,119 @@
+      % (100*lsvm.score(X_test, Y_test)))
+
+##############################################################################
+# Then we train linear SVMs on various numbers of :class:`PolynomialSampler`
+# features, showing that these features improve the accuracy of linear
+# classification:
```
I think it is more clear now.

lopeLH commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,119 @@
+
+for n_components in [200, 500, 1000, 1500]:
+
+    ps = PolynomialSampler(n_components=n_components,
+                           degree=4).fit(X_train[0:1])
+    X_train_ps = ps.transform(X_train)
+    X_test_ps = ps.transform(X_test)
+
+    lsvm = LinearSVC()
+    lsvm.fit(X_train_ps, Y_train)
```
- [ ] Use a Pipeline.

lopeLH commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,119 @@
+# to match the format of the dataset in the LIBSVM webpage [2].
+
+data = fetch_covtype()
+X, Y = data["data"], data["target"]
```
Fixed in the entire file.

lopeLH commented on sklearn/kernel_approximation.py:
```diff
@@ -22,7 +24,158 @@
 
 
+class PolynomialSampler(BaseEstimator, TransformerMixin):
+    """Implements Tensor Sketch [1], which approximates the feature map
```
fixed.

lopeLH commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,119 @@
+from sklearn.svm import SVC
+ksvm = SVC(C=500., kernel="poly", degree=4, coef0=0,
+           gamma=1.).fit(X_train, Y_train)
+print("Kernel-SVM score on raw features: %.2f %%" %
+      (100*ksvm.score(X_test, Y_test)))
```
done.

lopeLH commented on sklearn/kernel_approximation.py:
```diff
@@ -22,7 +24,158 @@
+    ----------
+    degree : int
+        Degree of the homogeneous polynomial kernel whose feature map
+        will be approximated. The kernel is K(x,y) = <x,y+coef0>^degree.
```
moved the formula to the docstring :+1:

lopeLH commented on sklearn/kernel_approximation.py:
```diff
@@ -22,7 +24,158 @@
+    def fit(self, X, y=None):
+        """Fit the model with X.
+
+        Initializes the internal variables. The method is data-independent,
```
avoided `data-independent` expression.

lopeLH commented on sklearn/kernel_approximation.py:
```diff
@@ -22,7 +24,158 @@
+        """Fit the model with X.
+
+        Initializes the internal variables. The method is data-independent,
+        so we only care about the dimension of samples in X.
```
Used `n_features` instead.

lopeLH commented on sklearn/kernel_approximation.py:
```diff
@@ -22,7 +24,158 @@
+        If RandomState instance, random_state is the random number generator;
+        If None, the random number generator is the RandomState instance used
+        by `np.random`.
+
```
done.

lopeLH commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,119 @@
+data = fetch_covtype()
+X, Y = data["data"], data["target"]
+
+Y[Y != 2] = 0
+Y[Y == 2] = 1
```
- [ ] Described data a bit more.

The reason for using `Y[Y == 2] = 1`, is because we try to reproduce one of the results from the original Tensor Sketch paper [1], where they used the [binary version of the Covertype dataset](https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary.html), as opposed to the standard 6-classes version. From the paper were this dataset was created [2]:

> We modified the 7-class classification problem into a binary classification problem where the goal was to separate class 2 from the other 6 classes.

So yeah, a bit of paper archeology. :man_scientist: 

[1] Pham, N., & Pagh, R. (2013, August). Fast and scalable polynomial kernels via explicit feature maps. In Proceedings of the 19th ACM SIGKDD international conference on Knowledge discovery and data mining (pp. 239-247)

[2] R. Collobert, S. Bengio, and Y. Bengio. A parallel mixture of SVMs for very large scale problems. Neural Computation, 14(05):1105-1114, 2002.

lopeLH commented on doc/modules/kernel_approximation.rst:
```diff
@@ -149,6 +149,28 @@ above for the :class:`RBFSampler`. The only difference is in the free
+
+The TensorSketch [PP2013]_ method, as implemented in :class:`PolynomialSampler`, is a
+scalable, data-independent method for polynomial kernel approximation. This method
+obtains a Count Sketch [CCF2002]_ of the outer product of two vectors (or a vector with
```
See my comment above.

rth replied:
Maybe 
"""
Count sketch is a dimensionality reduction technique similar to feature hashing, but that uses a several independent hash functions.
"""
to paraphrase https://en.wikipedia.org/wiki/Count_sketch Actually it could be useful to additionally have this link in the text. The original paper is good but it's still simpler to read on wiki for a casual reader.

lopeLH commented on doc/modules/kernel_approximation.rst:
```diff
@@ -149,6 +149,28 @@ above for the :class:`RBFSampler`. The only difference is in the free
+TensorSketch Method for Polynomial Kernel Approximation
+--------------------------------------------------------
+
+The TensorSketch [PP2013]_ method, as implemented in :class:`PolynomialSampler`, is a
```
I changed the title and described the poly-kernel first as requested. However, I am struggling a bit to describe `what a sketch is` in one sentence. I could explain how they are used in this algo, but that would be way longer... Do you have any suggestion?

lopeLH committed:
Address lorentzenchr's first round of comments.

lopeLH commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,119 @@
+# original Tensor Sketch paper [1].
+
+mm = MinMaxScaler().fit(X_train)
+X_train = normalize(mm.transform(X_train))
+X_test = normalize(mm.transform(X_test))
```
Again, trying to reproduce the same steps taken in the paper, which for normalization implies first min-max scaling to [0, 1], and then normalization of each sample to unit length.

lopeLH committed:
Apply suggestions from TomDLT's code review

Co-Authored-By: Tom Dupré la Tour <tom.dupre-la-tour@m4x.org>

lopeLH commented on sklearn/kernel_approximation.py:
```diff
@@ -22,7 +24,158 @@
+
+        self.bitHash_ = random_state.choice(a=[-1., 1.],
+                                            size=(self.degree, n_features)
+                                            ).astype(np.float32)
```
Not necessary indeed. Removed.

lopeLH committed:
Address TomDLT's first round of comments

lopeLH commented:
Thank you very much for your feedback @lorentzenchr @TomDLT. I think I addressed all your comments. There are some checks failing, but I believe these have nothing to do with this PR.

Regarding your question @lorentzenchr:

> It's always nice to have a plot. Is there something worth plotting in this case?

Well, I think there is nothing worth plotting in the example. It only produces a few accuracies. I do generate some plots in `benchmarks/bench_plot_polynomial_kernel_approximation`, which I pasted above.

TomDLT commented on sklearn/kernel_approximation.py:
```diff
@@ -22,7 +24,170 @@
+    >>> y = [0, 0, 1, 1]
+    >>> ps = PolynomialSampler(degree=3, random_state=1)
+    >>> X_features = ps.fit_transform(X)
+    &gt;&gt;&gt; clf = SGDClassifier(max_iter=10, tol=1e-3)
```
```suggestion
    >>> clf = SGDClassifier(max_iter=10, tol=1e-3)
```
I might have introduced this in my suggestion..

lopeLH committed:
Apply suggestions from TomDLT's code review

Co-Authored-By: Tom Dupré la Tour <tom.dupre-la-tour@m4x.org>

TomDLT commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,118 @@
+
+.. currentmodule:: sklearn.kernel_approximation
+
+We use the Covtype dataset [2]_, trying to reproduce the experiments on the
```
The links `[1]_`and `[2]_` do not work for some reason, not sure if they work in examples.
Maybe just use `[1]` and `[2]`.

lopeLH replied:
done.

TomDLT commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,118 @@
+y[y == 2] = 1  # We will try to separate class 2 from the other 6 classes.
+
+##############################################################################
+# Here we select 20,000 samples for training and 20,000 for testing.
```
The example takes 2 minutes to run, I don't know what is the rule for examples in scikit-learn but it seems a bit too long. Since the bottleneck should be the SVC, which scales poorly with the number of samples, we don't need to decrease the number of samples too much to largely decrease the computational time. Would the example still be meaningful with say 5000 or 10000 samples?

lopeLH replied:
Reduced the number training samples to `5,000`.

TomDLT commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,118 @@
+print(f"Kernel-SVM score on raw featrues: {ksvm_score:.2f} %%")
+
+##############################################################################
+# As we can see, this kernelized SVM achieves a slightly higher accuracy, but
```
I would be in favor of having a visual plot in this example, instead of having only printed accuracy.
I would picture for example an "accuracy versus time" plot, with a dot for each model.

lopeLH replied:
Okay, I modified the example a bit to generate the plot you suggested:
![image](https://user-images.githubusercontent.com/7440734/78471878-c5dfba80-7734-11ea-8179-0763b39feaae.png)

Of course the benefit of `PolynomialSampler` becomes more apparent when you have `100,000` training samples instead of `5,000`:

![image](https://user-images.githubusercontent.com/7440734/78471896-eb6cc400-7734-11ea-866d-e7f1aeef80d4.png)

But I think that is fine.

TomDLT commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,118 @@
+# reason why kernel approximators are so useful:
+
+from sklearn.svm import SVC
+ksvm = SVC(C=500., kernel="poly", degree=4, coef0=0,
```
Also, why do you use `C=500` here, while you use the default in `LinearSVC` ?

lopeLH replied:
The reason for that is that the experiment is designed so that if someone changes the number of training samples to `100,000`, we would reproduce the experiment in the original paper as closely as possible. Unfortunately, the paper does not provide the `C`-hyperparameters used, so I tried a few values and `C=1` in the `Linear SVM + PolynomialSampler` and `C=500` for the `Kernel SVM` seem to provide good results, which are also very close to the results reported in the paper. There is some discrepancy due to the stochastic nature of PolynomialSampler and the use of a random training/testing split. Some numbers:

```
Linear SVM with C=0.5 PolynomialSampler features: 79.16 %   -  t= 17.68 s
Linear SVM with C=1 PolynomialSampler features: 79.75 %   -  t= 21.98 s
Linear SVM with C=5 PolynomialSampler features: 80.01 %   -  t= 47.15 s
Linear SVM with C=6 PolynomialSampler features: 80.05 %   -  t= 51.63 s
---------------------------------------------------------------------------
Result in the original paper: 79.34%


Kernel-SVM with C=1 features: 79.91 %%   -  t= 521.89 s 
Kernel-SVM with C=5 features: 81.38 %%   -  t= 496.45 s
Kernel-SVM with C=50 features: 82.85 %%   -  t= 693.35 s
Kernel-SVM with C=100 features: 83.32 %%   -  t= 897.11 s
Kernel-SVM with C=500 features: 84.55 %%   -  t= 2343.52 s
Kernel-SVM with C=700 features: 84.60 %%   -  t= 3007.93 s
---------------------------------------------------------------------------
Result in the original paper: 84.01%
```

lopeLH committed:
Apply suggestions from TomDLT's code review

Co-Authored-By: Tom Dupré la Tour <tom.dupre-la-tour@m4x.org>

lopeLH committed:
Add plot to example

lopeLH commented:
@TomDLT @lorentzenchr , following you advice, I modified the example to produce a plot. Please have a look.

lorentzenchr commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,187 @@
+        pipeline = Pipeline(steps=[("kernel_approximator",
+                                    PolynomialSampler(
+                                        n_components=n_components,
+                                        degree=4)),
```
Can you say something about the `degree` in the example. With degree 4, there would be 54^4 ~ 8E6 output features. That's where `n_components` comes in.

lopeLH replied:
done!

lorentzenchr replied:
Sorry, I mixed up the polynomial kernel with `preprocessing.PolynomialFeatures`. It does not produce 54^4 output features.

lopeLH replied:
yeah, sure not! But the explicit kernel feature space would have indeed 54^4 features. What I added was:

> Note that, since the original samples have 54 features, the explicit feature map of the polynomial kernel of degree four would have 54^4 features. Thanks to :class:`PolynomialSampler`, we can condense most of the discriminative information of that feature space into a much more compact representation.

So the thing is that, to access the power of the kernel, you can 1) use a kernelized SVM, which uses the kernel function to access the feature space in an implicit manner, but has poor scalability. 2) Use the explicit kernel feature space, which in this case is unfeasible because it would have 54^4 features, or 3) Use a kernel approximation method such as PolynomialSampler, to condense the information of the kernel feature space to a compact representation.

Let me know if you would like to remove the note I added! It is accurate, but perhaps to the best place to mention this.

lorentzenchr replied:
Thanks for the explanation! So I was not so wrong after all:smile:
In this case, it is good to mention it and also give a number 54^4 ~ xxx. This way, the `n_components` becomes a better meaning.

lorentzenchr commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,187 @@
+
+##############################################################################
+# Train a kernelized SVM to see how well :class:`PolynomialSampler`
+# is approximating the performance of the kernel. This of course may take some
+# time, as the SVC class has a relatively poor scalability. This is the
```
```suggestion
# is approximating the performance of the kernel. This, of course, may take
# some time, as the SVC class has a relatively poor scalability. This is the
```

lorentzenchr commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,187 @@
+# References
+# ==========
+#
+# [1] Pham, N., & Pagh, R. (2013, August). Fast and scalable polynomial
+# kernels via explicit feature maps. In Proceedings of the 19th ACM SIGKDD
+# international conference on Knowledge discovery and data mining (pp. 239-247)
+# (https://chbrown.github.io/kdd-2013-usb/kdd/p239.pdf)
```
```suggestion
# [1] Pham, Ninh and Rasmus Pagh. "Fast and scalable polynomial kernels via
# explicit feature maps." KDD '13 (2013).
# https://doi.org/10.1145/2487575.2487591
```

lorentzenchr commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,187 @@
+# (https://chbrown.github.io/kdd-2013-usb/kdd/p239.pdf)
+#
+# [2] LIBSVM binary datasets repository
+# (https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary.html)
```
```suggestion
# https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary.html
```

lorentzenchr commented on sklearn/kernel_approximation.py:
```diff
@@ -22,7 +24,169 @@
+                        (iHashBit * X[:, j]).toarray().ravel()
+
+        else:
+            for j in range(X.shape[1]):
+                for d in range(self.degree):
+                    iHashIndex = self.indexHash_[d, j]
+                    iHashBit = self.bitHash_[d, j]
+                    count_sketches[:, d, iHashIndex] += iHashBit * X[:, j]
```
I tried to remove one loop (`j` or `d`) and do it vectorized. Vectorizing `d` like
```python
count_sketches = np.zeros((X.shape[0], 3, n_components))
for j in range(X.shape[1]):
    iHashIndex = self.indexHash_[:, j]
    iHashBit = self.bitHash_[:, j]
    count_sketches[:, np.arange(degree), iHashIndex] += iHashBit * X[:, j, np.newaxis]
```
was slower. Vectorizing `j` is tricky because `iHashIndex` can point to the same index for different values of `j`.
So let's keep it as is. Possible improvements maybe in later PRs.

lorentzenchr commented on sklearn/kernel_approximation.py:
```diff
@@ -22,7 +24,169 @@
+
+
 class RBFSampler(TransformerMixin, BaseEstimator):
+
```
```suggestion
```

lorentzenchr reviewed (commented):
Here comes another round.

lorentzenchr commented on sklearn/kernel_approximation.py:
```diff
@@ -22,7 +24,169 @@
+    n_components : int
+        Dimensionality of the output feature space.
+
+    random_state : int, RandomState instance or None, optional (default=None)
+        If int, random_state is the seed used by the random number generator.
+        If RandomState instance, random_state is the random number generator.
+        If None, the random number generator is the RandomState instance used
+        by `np.random`.
+
```
Better modify this to address #10548.

lopeLH replied:
done.

lopeLH committed:
Apply suggestions from lorentzenchr's second code review

Co-Authored-By: Christian Lorentzen <lorentzen.ch@googlemail.com>

lopeLH committed:
Add support for gamma hyperparameter of poly kernel

lopeLH commented on doc/modules/kernel_approximation.rst:
```diff
@@ -149,6 +149,44 @@ above for the :class:`RBFSampler`. The only difference is in the free
+
+.. math::
+
+        k(x, y) = (\gamma x^\top y +c_0)^d
```
It was not included in the original algorithm, but it is actually quite easy to support.

- [ ] Added support for gamma, and a test.

lorentzenchr commented:
@lopeLH I see you are working a lot on this PR very fast. It could help if you do not force push several times commits with the same subject, overwriting the last commit. Finish one commit at a time, then push, then work on the next commit. We are not in a rush:smirk:

lopeLH commented:
Ups, sorry, I was using github checks for debugging... But it is probably better to setup the testing and build scripts locally...

lorentzenchr commented:
At least for [PEP 8](https://www.python.org/dev/peps/pep-0008/) violations, it helps a lot to use an editor that supports a linter, e.g. with flake8 or pycodestyle.

lopeLH committed:
Fix doc error.

lopeLH committed:
Add small note about feature number to example comments

lopeLH committed:
Add gamma param to PolynomialSampler docstring kernel formula

lopeLH commented:
@lorentzenchr I believe I addressed all your second round comments :)

TomDLT commented on sklearn/kernel_approximation.py:
```diff
@@ -22,6 +24,167 @@
+    bitHash_ : ndarray of shape (degree, n_features), dtype=float32
+        Array with random entries in {+1, -1}, used to represent
+        the 2-wise independent hash functions for Count Sketch computation.
+    Examples
```
The tests are failing because you need an empty line here
```suggestion

    Examples
```

TomDLT reviewed (approved):
The example is much better with the plot, thanks !

TomDLT commented:
This new feature needs an entry in `doc/whats_new/v0.23.rst` !

lopeLH committed:
Apply suggestions from TomDLT's code review

Co-Authored-By: Tom Dupré la Tour <tom.dupre-la-tour@m4x.org>

lorentzenchr commented:
@lopeLH As a last point, the test could be simplified by something similar to
```python
@pytest.mark.parametrize('X', [X, csr_matrix(X)])
@pytest.mark.parametrize('Y', [Y, csr_matrix(Y)])
@pytest.mark.parametrize('gamma', [0.1, 1, 2.5])
@pytest.mark.parametrize('degree', [1, 2, 4])
@pytest.mark.parametrize('coef0', [0, 1, 4])
def test_polynomial_sampler(X, Y, gamma, degree, coef0):
    ...
    # compute exact kernel
    kernel = (gamma * X.dot(Y.T)+coef0)**degree
    ...
```
With a what's new entry, this looks good. Great work!

lopeLH committed:
Simplify test by using pytest.mark.parametrize

lopeLH committed:
Add what's new entry and fix linting error

lopeLH committed:
Fix one more linting error

lorentzenchr commented:
@lopeLH You can run your specific test locally via 
```
pytest -ra sklearn/tests/test_kernel_approximation.py::test_polynomial_sampler
```

lorentzenchr commented on sklearn/tests/test_kernel_approximation.py:
```diff
@@ -20,6 +21,33 @@
+    # kernel on random data
+
+    # compute exact kernel
+    kernel = (gamma * np.dot(X, Y.T) + coef0)**degree
```
Better use `metrics.pairwise.polynomial_kernel` as `np.dot` will fail for sparse matrices. And `polynomial_kernel` internally uses ` sklearn.utils.extmath.safe_sparse_dot` which takes care of the different cases.

lopeLH replied:
Done, thank you very much for the tips!

lopeLH committed:
Fix test on sparse data

lopeLH commented:
Thank you very much @lorentzenchr @TomDLT  for your feedback, advice and patience :)

lorentzenchr commented:
@lopeLH As you already have one core dev's approval, you can change the title to [MRG+1].

lorentzenchr commented:
@lopeLH It seems that [SLEP009](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep009/proposal.html) will be implemented with the next upcoming release. So this PR should make the arguments of `PolynomialSampler.__init__` keyword only, see #16474.

lopeLH committed:
Make init arguments keyword-only

lopeLH committed:
Reduce n_runs in example to avoid hittin CI runtime limit

lopeLH committed:
Update v0.23.rst whatsnew entry

lopeLH committed:
Merge branch 'master' into add-tensorSketch

lopeLH committed:
Trigger CI

lopeLH commented:
@lorentzenchr Do you have any clue why the `LGTM analysis: Python` check is failing, and how to fix it?

lorentzenchr commented:
There is something wrong with the git status of this branch. In LGTM, is says
> You are in 'detached HEAD' state.

When I try to delete this branch locally, `git branch -d add-tensorSketch`, is says
> error: The branch 'add-tensorSketch' is not fully merged.

@TomDLT @GaelVaroquaux Can you help out? I'm not a git guru.

rth commented:
> There is something wrong with the git status of this branch. In LGTM, is says You are in 'detached HEAD' state.

The detached HEAD mode might be normal depending on what LGTM.com is doing (e.g. it happens when one checks out a commit that's not at the edge of a branch). I cannot see any actual errors in LGTM.  Feel free to ignore the LGTM check it shouldn't prevent merging this PR. Maybe merging master in again could help. @lopeLH In general, I would recommend not force pushing, but just adding new commits, as we squash at the end.

lopeLH committed:
Merge pull request #1 from scikit-learn/master

Merge master

lopeLH committed:
Merge branch 'master' into add-tensorSketch

lopeLH commented:
Thanks for the advice @rth, sorry about the force pushing, It was me trying to fix what caused LGTM to fail. I merged master into this branch again, but the check continues failing, so I will give up on it for now, as you suggested.

lorentzenchr commented:
Friendly ping at core devs, in particular @amueller @GaelVaroquaux @agramfort. This PR is in a very good shape and already has one +1 (if you somehow count mine, it has +1.5:smirk:). It would be great to finalize it.

@lopeLH Could you move the what's new entry to v0.24?

lopeLH committed:
Move entry from v0.23.rst to v0.24.rst

lopeLH committed:
Merge branch 'master' into add-tensorSketch

lopeLH committed:
Fix v0.24.rst

lopeLH committed:
Add whatsnew entry it v0.24.rst

lopeLH committed:
Fix linting error (F541)

cmarmo commented:
Hi @lopeLH do you mind fixing conflicts (again... sorry)? Thanks for your patience!
@agramfort perhaps you have some time to have a look? Thanks! As @lorentzenchr said it would be great to finalize this one.

lopeLH committed:
Merge branch 'master' into add-tensorSketch

lopeLH commented:
No problem @cmarmo, done!

lopeLH committed:
Merge branch 'master' into add-tensorSketch

rth commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,191 @@
+# classification problem to match the version of the dataset in the
+# LIBSVM webpage [2], which was the one used in [1].
+
+data = fetch_covtype()
+X, y = data["data"], data["target"]
```
```suggestion
X, y = fetch_covtype(return_X_y=True)
```

rth commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,191 @@
+# the LIBSVM webpage, and then normalize to unit length as done in the
+# original Tensor Sketch paper [1].
+
+mm = MinMaxScaler().fit(X_train)
+X_train = normalize(mm.transform(X_train))
+X_test = normalize(mm.transform(X_test))
```
Then let's use a pipeline.

```suggestion
mm = make_pipeline(MinMaxScaler(), Normalizer())
X_train = mm.fit_transform(X_train)
X_test = mm.transform(X_test)
```

rth commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,191 @@
+    ps_lsvm_time /= n_runs
+    ps_lsvm_score /= n_runs
+
+    results[f"LSVM + PS({n_components})"] = \
+        {"time": ps_lsvm_time, "score": ps_lsvm_score}
```
We try to avoid `\` unless there is no choice,
```suggestion
    results[f"LSVM + PS({n_components})"] = {
        "time": ps_lsvm_time, "score": ps_lsvm_score
    }
```

rth commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,191 @@
+# but its training time is much larger and, most importantly, will grow
+# much faster if the number of training samples increases.
+
+plt.figure(figsize=(7, 7))
+plt.scatter([results["LSVM"]["time"], ], [results["LSVM"]["score"], ],
+            label="Linear SVM", c="green", marker="^")
+
+plt.scatter([results["LSVM + PS(250)"]["time"], ],
+            [results["LSVM + PS(250)"]["score"], ],
+            label="Linear SVM + PolynomialSampler", c="blue")
+plt.scatter([results["LSVM + PS(500)"]["time"], ],
+            [results["LSVM + PS(500)"]["score"], ],
+            c="blue")
+plt.scatter([results["LSVM + PS(1000)"]["time"], ],
+            [results["LSVM + PS(1000)"]["score"], ],
+            c="blue")
+plt.scatter([results["LSVM + PS(2000)"]["time"], ],
+            [results["LSVM + PS(2000)"]["score"], ],
+            c="blue")
+plt.annotate("n_comp.=250", (results["LSVM + PS(250)"]["time"],
+                             results["LSVM + PS(250)"]["score"]),
+             xytext=(-30, 10), textcoords="offset pixels")
+plt.annotate("n_comp.=500", (results["LSVM + PS(500)"]["time"],
+                             results["LSVM + PS(500)"]["score"]),
+             xytext=(-30, 10), textcoords="offset pixels")
+plt.annotate("n_comp.=1000", (results["LSVM + PS(1000)"]["time"],
+                              results["LSVM + PS(1000)"]["score"]),
+             xytext=(-30, 10), textcoords="offset pixels")
+plt.annotate("n_comp.=2000", (results["LSVM + PS(2000)"]["time"],
+                              results["LSVM + PS(2000)"]["score"]),
+             xytext=(-30, 10), textcoords="offset pixels")
+
+plt.scatter([results["KSVM"]["time"], ], [results["KSVM"]["score"], ],
+            label="Kernel SVM", c="red", marker="x")
+
+plt.xlabel("Training time (s)")
+plt.ylabel("Accurary (%)")
+plt.legend()
+plt.show()
```
The matplotlib object API is preffered as it makes it easier to manipulate individual figures https://matplotlib.org/3.3.0/api/index.html#usage-patterns

Also some parametrization,

```suggestion
fig, ax = plt.subplots(figsize=(7, 7))
ax.scatter([results["LSVM"]["time"], ], [results["LSVM"]["score"], ],
            label="Linear SVM", c="green", marker="^")

ax.scatter([results["LSVM + PS(250)"]["time"], ],
            [results["LSVM + PS(250)"]["score"], ],
            label="Linear SVM + PolynomialSampler", c="blue")
for n_components in [250, 500, 1000, 2000]:  # preferably define N_COMPONENTS above
    ax.scatter([results[f"LSVM + PS({n_components})"]["time"], ],
               [results[f"LSVM + PS({n_components})"]["score"], ],
               c="blue")
	ax.annotate(f"n_comp.={n_components}",
				 (results[f"LSVM + PS({n_components})"]["time"],
                  results[f"LSVM + PS({n_components})"]["score"]),
            	 xytext=(-30, 10), textcoords="offset pixels")

ax.scatter([results["KSVM"]["time"], ], [results["KSVM"]["score"], ],
            label="Kernel SVM", c="red", marker="x")

ax.set_xlabel("Training time (s)")
ax.set_ylabel("Accurary (%)")
ax.legend()
plt.show()
```

rth commented on sklearn/kernel_approximation.py:
```diff
@@ -1,17 +1,19 @@
 import numpy as np
 import scipy.sparse as sp
 from scipy.linalg import svd
+from scipy import fftpack
```
fftpack is legacy, it should be `scipy.fft` on newer scipy versions https://docs.scipy.org/doc/scipy/reference/fftpack.html

Though might have to still use fftpack on older scipy versions.

lopeLH replied:
I changed that, but `doc-min-dependencies` failed. What should I do?

https://app.circleci.com/pipelines/github/scikit-learn/scikit-learn/6636/workflows/80111186-c49a-4a6c-808d-6c222566c631/jobs/115998

lopeLH replied:
Sadly `scipy.fft` is a function you can import even in `scipy <= 1.3`, but the following should work:
```
try:
    from scipy.fft import fft, ifft
except ImportError:   # scipy < 1.4
    from scipy.fftpack import fft, ifft
```

rth commented on sklearn/kernel_approximation.py:
```diff
@@ -22,6 +24,168 @@
+        K(X, Y) = (gamma * <X, Y> + coef0)^degree
+
+    by efficiently computing a Count Sketch of the outer product of a
+    vector with itself. Read more in the
```
Maybe,
```suggestion
    vector with itself using Fast Fourier Transforms (FFT). Read more in the
```
to be a bit more concrete.

rth commented on sklearn/kernel_approximation.py:
```diff
@@ -22,6 +24,168 @@
+        Parameter of the polynomial kernel whose feature map
+        will be approximated.
+
+    degree : int
```
```suggestion
    degree : int, default=2
```

rth commented on sklearn/kernel_approximation.py:
```diff
@@ -22,6 +24,168 @@
+
+    Parameters
+    ----------
+    gamma : float
```
```suggestion
    gamma : float, default=1.0
```

rth commented on sklearn/kernel_approximation.py:
```diff
@@ -22,6 +24,168 @@
+        Degree of the polynomial kernel whose feature map
+        will be approximated.
+
+    coef0 : int
```
```suggestion
    coef0 : int, default=0
```

rth commented on sklearn/kernel_approximation.py:
```diff
@@ -22,6 +24,168 @@
+        Constant term of the polynomial kernel whose feature map
+        will be approximated.
+
+    n_components : int
```
```suggestion
    n_components : int, default=100
```

rth commented on sklearn/kernel_approximation.py:
```diff
@@ -22,6 +24,168 @@
+            Returns the transformer.
+        """
+
+        X = check_array(X, accept_sparse="csc")
```
This should use 
```suggestion
        X = self._validate_data(X, accept_sparse="csc")
```
which calls check_array internally,

rth commented on sklearn/kernel_approximation.py:
```diff
@@ -22,6 +24,168 @@
+        """
+
+        X = check_array(X, accept_sparse="csc")
+        random_state = check_random_state(self.random_state)
```
The issue is that calling fit twice would produce a different seed and therefore a different result, since RandomState instance is mutable. In https://github.com/scikit-learn/scikit-learn/pull/14605 for a similar use case, we added a `seed` variable in transform, but I'm not particularly happy with that outcome either. This is probably fine as is, we would just have to address this globally at some point in https://github.com/scikit-learn/scikit-learn/issues/14042

rth commented on sklearn/kernel_approximation.py:
```diff
@@ -22,6 +24,168 @@
+        """
+
+        check_is_fitted(self)
+        X = check_array(X, accept_sparse="csc")
```
```suggestion
        X = self._validate_data(X, accept_sparse="csc")
```

rth commented on sklearn/kernel_approximation.py:
```diff
@@ -22,6 +24,168 @@
+
+        X_gamma = np.sqrt(self.gamma) * X
+
+        if sp.issparse(X) and self.coef0 != 0:
```
```suggestion
        if sp.issparse(X_gamma) and self.coef0 != 0:
```
to be consistent with the check in the second branch

rth commented on sklearn/kernel_approximation.py:
```diff
@@ -22,6 +24,168 @@
+                                         overwrite_x=True)
+        count_sketches_fft_prod = np.prod(count_sketches_fft, axis=1)
+        data_sketch = np.real(fftpack.ifft(count_sketches_fft_prod,
+                                           overwrite_x=True))
```
This can be a follow up issue/PR, and would need double checking but since the count_sketches input is real you can likely use `rfft` and `rfft` witch would be faster.

lopeLH replied:
Wrote that down so we don't forget to have a look

rth commented on benchmarks/bench_plot_polynomial_kernel_approximation.py:
```diff
@@ -0,0 +1,155 @@
+
+# Evaluate PolynomialSampler + LinearSVM
+ps_svm_scores = []
+n_test = 5
```
Better to call this `n_runs` as  "test" is a already quite overloaded term in CS and ML contexts.

lopeLH replied:
Fixed.

rth reviewed (commented):
Thanks for your work and patience. Also thanks for reviews!  We should certainly include this in the next release. 

A few more comments below but otherwise it looks good.

rth committed:
Merge branch 'master' into add-tensorSketch

TomDLT commented on benchmarks/bench_plot_polynomial_kernel_approximation.py:
```diff
@@ -0,0 +1,155 @@
+    ny_svm_scores.append(100*score_avg/n_test)
+
+# Show results
+plt.figure(figsize=(6, 4))
+plt.title("Accuracy results")
+plt.plot(out_dims, ps_svm_scores, label="PolynomialSampler + linear SVM",
+         c="orange")
+plt.plot(out_dims, ny_svm_scores, label="Nystroem + linear SVM",
+         c="blue")
+plt.plot([out_dims[0], out_dims[-1]], [lsvm_score, lsvm_score],
+         label="Linear SVM", c="black", dashes=[2, 2])
+plt.plot([out_dims[0], out_dims[-1]], [ksvm_score, ksvm_score],
+         label="Poly-kernel SVM", c="red", dashes=[2, 2])
+plt.legend()
+plt.xlabel("N_components for PolynomialSampler and Nystroem")
+plt.ylabel("Accuracy (%)")
+plt.xlim([out_dims[0], out_dims[-1]])
+plt.tight_layout()
```
`fig.tight_layout` ?

rth replied:
Indeed, thanks. Updated. We probably should avoid tight_layout in any case in favor of [contrained_layout](https://matplotlib.org/3.1.3/tutorials/intermediate/constrainedlayout_guide.html) but it's likely not yet supported by the minimal matplotlib version used for examples..

lorentzenchr commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,191 @@
+from sklearn.pipeline import Pipeline
+import time
+
+##############################################################################
```
New block separator `# %%`instead of `#######`, see #17063 and # 17068.
```suggestion
# %%
```

lorentzenchr commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,191 @@
+y[y != 2] = 0
+y[y == 2] = 1  # We will try to separate class 2 from the other 6 classes.
+
+##############################################################################
```
```suggestion
# %%
```

lorentzenchr commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,191 @@
+                                                    test_size=10_000,
+                                                    random_state=42)
+
+##############################################################################
```
```suggestion
# %%
```

lorentzenchr commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,191 @@
+X_test = normalize(mm.transform(X_test))
+
+
+##############################################################################
```
```suggestion
# %%
```

lorentzenchr commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,191 @@
+results["LSVM"] = {"time": lsvm_time, "score": lsvm_score}
+print(f"Linear SVM score on raw features: {lsvm_score:.2f}%")
+
+##############################################################################
```
```suggestion
# %%
```

lorentzenchr commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,191 @@
+    print(f"Linear SVM score on {n_components} PolynomialSampler " +
+          f"features: {ps_lsvm_score:.2f}%")
+
+##############################################################################
```
```suggestion
# %%
```

lorentzenchr commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,191 @@
+results["KSVM"] = {"time": ksvm_time, "score": ksvm_score}
+print(f"Kernel-SVM score on raw featrues: {ksvm_score:.2f}%")
+
+##############################################################################
```
```suggestion
# %%
```

lorentzenchr commented on examples/plot_scalable_poly_kernels.py:
```diff
@@ -0,0 +1,191 @@
+plt.legend()
+plt.show()
+
+##############################################################################
```
Here I'm not quite sure.
```suggestion
# %%
```

lopeLH committed:
Make y / y_train / y_test lowercase to indicate vector

lopeLH commented on benchmarks/bench_plot_polynomial_kernel_approximation.py:
```diff
@@ -0,0 +1,155 @@
+from sklearn.pipeline import Pipeline
+
+# Split data in train and test sets
+X, Y = load_digits()["data"], load_digits()["target"]
```
Fixed.

lopeLH committed:
Merge branch 'add-tensorSketch' of github.com:lopeLH/scikit-learn into add-tensorSketch

lopeLH committed:
Update benchmarks/bench_plot_polynomial_kernel_approximation.py

Use object matplotlib API

Co-authored-by: Roman Yurchak <rth.yurchak@gmail.com>

lopeLH committed:
Update benchmarks/bench_plot_polynomial_kernel_approximation.py

Use object matplotlib API x2

Co-authored-by: Roman Yurchak <rth.yurchak@gmail.com>

lopeLH committed:
Fix linting error

lopeLH committed:
Add wikipedia page reference for Count Sketch

lopeLH committed:
Update examples/plot_scalable_poly_kernels.py

Use return_X_y=True when fetching data

Co-authored-by: Roman Yurchak <rth.yurchak@gmail.com>

lopeLH committed:
Update examples/plot_scalable_poly_kernels.py

Use a pipeline when chaining preproc steps

Co-authored-by: Roman Yurchak <rth.yurchak@gmail.com>

lopeLH committed:
Update sklearn/kernel_approximation.py

Use _validate_data

Co-authored-by: Roman Yurchak <rth.yurchak@gmail.com>

lopeLH committed:
Update sklearn/kernel_approximation.py

Co-authored-by: Roman Yurchak <rth.yurchak@gmail.com>

lopeLH committed:
rename n_tests by n_runs

lopeLH committed:
Add missing imports

lopeLH committed:
Use new separator format: # %%

Co-authored-by: Christian Lorentzen <lorentzen.ch@googlemail.com>

lopeLH committed:
Avoid using \

Co-authored-by: Roman Yurchak <rth.yurchak@gmail.com>

lopeLH committed:
Use object matplotlib API

Co-authored-by: Roman Yurchak <rth.yurchak@gmail.com>

lopeLH committed:
Improved comment

Co-authored-by: Roman Yurchak <rth.yurchak@gmail.com>

lopeLH committed:
Use _validate_data

Co-authored-by: Roman Yurchak <rth.yurchak@gmail.com>

lopeLH committed:
Add default values to docstrings

Co-authored-by: Roman Yurchak <rth.yurchak@gmail.com>

lopeLH committed:
Fix linting error

lopeLH committed:
Define N_COMPONENTS above. Linting error.

lopeLH committed:
Use scipy.fft instead of legacy scipy.fftpack

lopeLH committed:
Back to fftpack

lopeLH committed:
Fix reference in docs

rth commented:
> Back to fftpack 

You could use,
```py
try:
    from scipy import fft
except ImportError:   # scipy < 1.4
   from scipy import fftpack as fft
```
the scipy.fft uses their newer pocketfft implementation https://github.com/scipy/scipy/pull/10238 that's more optimized https://github.com/numpy/numpy/pull/11888#issuecomment-419172698

lopeLH committed:
Use scipy.fft if available, otherwise revert to scipy.fftpack

lopeLH committed:
Update examples/plot_scalable_poly_kernels.py

Co-authored-by: Christian Lorentzen <lorentzen.ch@googlemail.com>

lopeLH committed:
Move references of Count sketch to the first mention of this method

lopeLH commented:
@rth I think I addressed all your comments. All checks passing :)

Maybe have a look at how I phrased things at `doc/modules/kernel_approximation.rst` (line 175), regarding Count sketch and its role in TensorSketch.

rth commented:
Thanks! I have been experimenting with using this approach for text classification on a 10k samples subset of AG News dataset. Granted it's probably not a typical use case, but I still wanted to check that results generally make sense. Below are classification accuracy results,

| label                                                        |   fit_time |   train accuracy |   test accuracy |
|:-------------------------------------------------------------|-----------:|-----------------:|----------------:|
| baseline: CountVectorizer + LinearSVC                        |   0.18 |         0.98   |          0.84 |
| baseline w/ PCA(n_components=100)                            |   0.64 |         0.84 |          0.83 |
| baseline w/ SparseRandomProjection(n_components=1000)        |   0.29 |         0.51   |          0.49 |
| baseline w/ SVC(kernel='poly', degree=2)                     |   9.68  |         0.99 |          0.86 |
| baseline w/ PolynomialSampler(degree=2, n_components=100)    |   1.12  |         0.36 |          0.29 |
| baseline w/ PolynomialSampler(degree=2, n_components=1000)   |   3.94  |         0.65 |          0.43 |
| baseline w/ PolynomialSampler(degree=2, n_components=10000)  |  25.34    |         0.99   |          0.62 |
| baseline w/ PolynomialSampler(degree=2, n_components=20000)  |  53.13   |         0.99    |          0.72 |
| baseline w/ PolynomialSampler(degree=2, n_components=40000)  | 106.51    |         0.99   |          0.78 |
| baseline w/ PolynomialSampler(degree=2, n_components=100000) | 269.17    |         0.99 |          0.84 |

obtained with the following notebook [tensorsketch-experiments-sparse.py](https://github.com/scikit-learn/scikit-learn/files/5068436/tensorsketch-experiments-sparse.py.txt). Here test scores are done without cross-validation as it takes already a long time (and requires the above optimization for sparse), so they are not too reliable. There are also likely overfitting issues with linearSVC and a large number of components.

Main takeaways,
  1. we need to explicitly check in fit that `degree>=1` otherwise FFT fails with a 0D input matrix for degree=0. It would be also good to add a test that check this exception with `with pytest.raises(ValueError, match='degree=0 should be >=1.')`.
  2. results will be very bad for  n_components < n_features even evaluating on the train subset. As far as I can tell, both the paper and the example only illustrate the case n_components > n_features, with an optimum of evaluation score / run time cost around n_components=10*n_features. We should adds some of these suggestions in the docstring of `n_components` (and add a sentence to the user manual) to help choose n_components. Otherwise users may be very disappointing with the performance when using the default `n_components=100` on a higher dimensional case. 
  3. For the sparse case 2. implies that it's not very usable, because one gets as output dense matrices with 10k+ features. Empirically, with sparse input and for  n_components > n_features, the resulting dense matrix actually contains mostly zeros. So in a follow up PR, it could be worth considering whether it would make sense thesholding (in the above example 1e-15 would be ok as a threshold) and converting back to sparse. Though the intermediary step with the FFT would still be dense with the associated memory requirements, maybe it could be worth chunking with respect to n_samples not sure.

Overall +1 for merging after a few more documentation precisions on how to choose `n_components` and a check/test for `degree<1`.

rth commented:
Another comment for reviewers is that I'm not convinced by the name `PolynomialSampler`. This algorithm does no sampling as far as I can see, and there is 0 occurrence of word sampling in ether of cited papers. Rather, roughly, it applies a convolution on hashed feature spaces (via product in FFT space). Also the actual implementation looks more like a random projection than hashing. So the current name is a bit misleading IMO.

The original name `TensorSketch` might be a bit opaque but how about,
 - `PolynomialCountSketch`or `PolynomialTensorSketch`: the count sketch is the correct name for this hashing technique. It may be a bit exotic/new (invented in 2003) but there is a wiki page https://en.wikipedia.org/wiki/Count_sketch and around 14k search results
 -  `PolynomialFeatureHasher` could have worked except that `FeatureHasher` works with completely different input
 - `PolynomialProjection` by analogy with random projection as it's a bit related
 - or some name with a combination of "Polynomial" and "Hashing" or "Projection"

WDYT?

lopeLH committed:
Check and test degree >= 1

lopeLH committed:
Use ValueError instead of AssertionError

lopeLH committed:
Test raises if degree<=0 for various degrees

lopeLH committed:
Style edit

lopeLH committed:
Remove leftover comment

lopeLH committed:
Added recomendations for n_components in docstring and user-guide

lopeLH commented:
@rth, as you requested:

- [x] Added a check enforcing `degree >= 1` in the `__init__` of `PolynomialSampler`.
- [x] Added a test checking `PolynomialSampler` throws if initialized with a `degree` lower than one.
- [x] Added the suggested hints regarding the selection of `n_components` to the docstring and user guide (please, have a look, I don't trust my English).

Regarding the name of the main class, `PolynomialCountSketch` sounds good to me.

TomDLT commented on sklearn/kernel_approximation.py:
```diff
@@ -22,6 +27,173 @@
+    def __init__(self, *, gamma=1., degree=2, coef0=0, n_components=100,
+                 random_state=None):
+
+        if not degree >= 1:
```
The check should not be in the __init__ ([cf doc](https://scikit-learn.org/stable/developers/develop.html?highlight=init__#parameters-and-init)).

lopeLH replied:
Sorry about that. Fixed.

TomDLT reviewed (commented):
`PolynomialCountSketch` sounds better indeed.

lopeLH committed:
Move degree check to fit method

rth reviewed (approved):
Thanks, LGTM. Let's wait a few more days to see if there are any objection to the `PolynomialCountSketch` name (cf https://github.com/scikit-learn/scikit-learn/pull/13003#issuecomment-673430807). And if there are none, rename it and merge.

lorentzenchr commented:
+1 for `PolynomialCountSketch` from my side.

lopeLH committed:
Rename main class to PolynomialCountSketch
(and hope nothing breaks)

lopeLH commented:
Seems like everyone is happy with the new name (`PolynomialCountSketch`), so I performed the change. 

Do I have to squash the ugly commit history in this branch into a single, clean commit? Anyways, let me know if there is anything left on my side.

Super excited about having my first contribution to sklearn merged! :partying_face:

lorentzenchr commented:
@lopeLH Thank you very much for your contribution! And feel free to continue - if you'd like on one of the possible follow-ups that @rth listed.
You don't need to squash commits. That's done automatically when we merge.

I'm also excited as this is my first merge hoping everything went fine.

GaelVaroquaux commented:
Woohoo! Great! Exciting!
